### PR TITLE
Fix type test for trait parameter arguments

### LINKED ISSUE #11993: Failure to infer dependently typed constructor parameter (regression)
## Compiler version

3.0.0-RC2

## Minimized code

```Scala
class Foo(
  val x: String,
  val y: Option[x.type]
)

class Bar extends Foo("bar", Some("bar"))
```

## Output

In 3.0.0-RC1, this compiles, as expected.

In 3.0.0-RC2:

```scala
6 |class Bar extends Foo("bar", Some("bar"))
  |      ^
  |illegal parameter: The types of value y do not match.
  |
  |  value y in class Foo has type: Option[(Foo#x : String)]
  |  but class Bar expects value y to have type: Option[(Bar.this.x : String)]
1 error found

```

### PR DESCRIPTION
**Pull request by odersky**
Fixes #11993

### CODE REVIEW

**Comment by odersky** [comment]
I have added this to the 3.0 milestone, so that it is backported to 3.0 RC3. I think the previous implementation was too restrictive and risks disallowing various code that is perfectly legal. The previous failure in #3989 is more concerning in that respect than #11993.

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 1]
```diff
@@ -133,6 +133,27 @@ object RefChecks {
...
+       *  in the current class. This is necessary since parameter types might be narrowed
+       *  through intersection with other parent traits. See neg/i11018.scala.
+       */
+      def checkParamInits(app: Apply): Unit =
+        val parentCls = app.tpe.classSymbol
+        if parentCls.is(Trait) then
+          val params = parentCls.asClass.paramGetters
+          val args = termArgss(app).flatten
+          for (param, arg) <- params.lazyZip(args) do
+            if !param.is(Private) then // its type can be narrowed through intersection -> a check is needed
```
Comment (b-studios) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:146]: It is a good idea to check this one here, I didn't do that in #11059.

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 2]
```diff
@@ -133,6 +133,27 @@ object RefChecks {
...
+              val paramType = cls.thisType.memberInfo(param)
+              if !(arg.tpe <:< paramType) then
+                val argTypes = args.tpes
+                // it could still be OK but we might need to substitute arguments for parameters
+                // to account for dependent parameters. See pos/i11993.scala
+                if !(arg.tpe.subst(params, argTypes) <:< paramType.subst(params, argTypes))
+                then
+                  report.error(IllegalParameterInit(arg.tpe, paramType, param, cls), arg.srcPos)
+
+      for case app: Apply <- parentTrees do checkParamInits(app)
```
Comment (b-studios) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:156]: yes, it's also a better idea to walk through the constructors.

**Review by b-studios — APPROVED** [review]
LGTM much more thought out than my #11059

**Pull request merged**